src/hb-common.cc

```diff
@@ -665,6 +666,125 @@ hb_script_get_horizontal_direction (hb_script_t script)
   return HB_DIRECTION_LTR;
 }
 
+static int
+lang_compare_first_component (const void *pa,
+                              const void *pb)
+{
+  const char *a = (const char *) pa;
+  const char *b = (const char *) pb;
+  unsigned int da, db;
+  const char *p;
+
+  p = strstr (a, "-");
+  da = p ? (unsigned int) (p - a) : strlen (a);
+
+  p = strstr (b, "-");
+  db = p ? (unsigned int) (p - b) : strlen (b);
+
+  return strncmp (a, b, hb_max (da, db));
+}
+
+static const HbScriptForLang *
+find_best_lang_match (hb_language_t language)
+{
+  const char *lang_str;
+  const char *record, *start, *end;
+  const char *records;
+  unsigned int record_size;
+  unsigned int num_records;
+
+  if (language == nullptr)
+    return nullptr;
+
+  lang_str = language->s;
+
+  records = (const char *) hb_script_for_lang;
+  record_size = sizeof (HbScriptForLang);
+  num_records = sizeof (hb_script_for_lang) / record_size;
+  record = (const char *) bsearch (lang_str,
+                                   records, num_records,
+                                   record_size,
+                                   lang_compare_first_component);
+  if (!record)
+    return nullptr;
+
+  start = records;
+  end   = start + num_records * record_size;
+
+  /* find the best match among all those that have the same first-component */
+
+  /* go to the final one matching in the first component */
+  while (record < end - record_size &&
+         lang_compare_first_component (lang_str, record + record_size) == 0)
+
+  /* go back, find which one matches completely */
+  while (start <= record &&
+         lang_compare_first_component (lang_str, record) == 0)
+    {
+      hb_language_t l;
+
+      l = hb_language_from_string (record, -1);
+      if (hb_language_matches (language, l))
+        return (const HbScriptForLang *) record;
+
+      record -= record_size;
+    }
+
+  return nullptr;
+}
+
+/**
+ * hb_language_get_scripts:
+ * @language: a #hb_language_t
+ * @script_count: (inout) (optional): Input = the maximum number of scripts to return;
+ *      Output = the actual number of scripts returned (may be zero)
+ * @scripts: (out) (array length=script_count): the array of #hb_script_t found
+ *
+ * Fetches the scripts that can be used with @language.
```

Comment:
This is a bit vague, since languages can practically be written in any any given script. Also I’m not sure what this API is useful for.

Reply:
> This is a bit vague, since languages can practically be written in any any given script. Also I’m not sure what this API is useful for.

My immediate use is finding a direction from a language.

Reply:
I don’t follow. Direction is a property of the characters (i.e. from UCD and BiDi algorithm), when does language play a role?

Reply:
Pango goes from language tag -> script -> direction with this API.

Reply:
When you show a UI in a given locale, and you need to decide whether to put the sidebar on the left or on the right.

Reply:
For the main layout direction of a language. We have `hb_script_get_horizontal_direction`.

Reply:
you could of course cut out the script here, and just provide a hb_language_get_horizontal_direction()

Reply:
That sounds like a rather odd way to get the direction of a locale (and locales != languages). IIRC, GTK had a special string you translate to LTR or RTL to set the UI direction for a given translation. But also it feels very pango-speficic and not of much value to anyone else and I don’t see why Pango can’t keep using the code it has even after switching to `hb_language_t` as it does seem to be using any insider HarfBuzz knowledge about the langues here.

Reply:
We want to get out of abusing translations for this. Not the least because it turns out that random other projects open gtk.mo just to pull out the translation for default:LTR

Reply:
And I want to get rid of the code in pango, because I am the only one who ever touches pango, whereas there are multiple contributors here.

Reply:
Yeah I find language<->script to be generally useful, plus also script_get_prominent_language(). These, of course as Khaled says, don't have any use in HarfBuzz itself, but are useful to higher level specially to itemization and font selection layers.

Reply:
I don’t know, but language feels like the wrong level for this information to me. Besides, what does this API return for, say, `ms-Arab` or `sr-Latn`?

Reply:
Our language tags *are* BCP 47. So we can make it understand those.

Reply:
I still don’t like this, but at least give it some less-confident name, like `hb_langauge_guess_script()` or some such and make it clear in the documentation this is a best guess (not that it is going to stop any one, people still use `hb_buffer_guess_segemet_properties()` in lieu of proper segmentation).

Reply:
There is also [uppress-Script Field](https://www.rfc-editor.org/rfc/rfc5646.html#section-3.1.9) of the [IANA Language Subtag Registry](http://www.iana.org/assignments/language-subtag-registry/language-subtag-registry). At this point, one might as well use https://github.com/tagoh/liblangtag.

Reply:
Other use cases for language <> script relation that I find: Pango gets text that is marked up with a language, and wants to ensure that the language matches the script of the text, to avoid nonsensical situations.

Reply:
> I don’t know, but language feels like the wrong level for this information to me.

You mean the direction, here?

Reply:
> Other use cases for language <> script relation that I find: Pango gets text that is marked up with a language, and wants to ensure that the language matches the script of the text, to avoid nonsensical situations.

But is this necessary? Many languages are written in scripts they are not predominantly written in (either historically or by minority or diaspora communities), what languages should these texts be tagged in. What kind of issues Pango is preventing by filtering out unlikely languages? Shaping depends on the script and language influence is usually minimal.

Reply:
> > I don’t know, but language feels like the wrong level for this information to me.
> 
> You mean the direction, here?

I mean getting direction from the language is the wrong level.

Reply:
> But is this necessary? 

I don't know. This is old code that I did not write:

```
static PangoLanguage *
compute_derived_language (PangoLanguage *lang,
                          PangoScript    script)
{
  PangoLanguage *derived_lang;

  /* Make sure the language tag is consistent with the derived
   * script. There is no point in marking up a section of
   * Arabic text with the "en" language tag.
   */
  if (lang && pango_language_includes_script (lang, script))
    derived_lang = lang;
  else
    {
      derived_lang = pango_script_get_sample_language (script);
      /* If we don't find a sample language for the script, we
       * use a language tag that shouldn't actually be used
       * anywhere. This keeps fontconfig (for the PangoFc*
       * backend) from using the language tag to affect the
       * sort order. I don't have a reference for 'xx' being
       * safe here, though Keith Packard claims it is.
       */
      if (!derived_lang)
        derived_lang = pango_language_from_string ("xx");
    }

  return derived_lang;
}
```

Reply:
IMHO, that feels like a solution in search of a problem.

Reply:
Reading https://bugzilla.gnome.org/show_bug.cgi?id=123581, it seems the root of the issue is that Pango defaults to the locale language which might be pointless but harmless, except I guess for FontConfig, but I don’t know what FontConfig does with the language so I don’t know how bad is it. Also it is not making a distinction between explicitly set language (which should be unconditionally taken as is) and the locale-driven one. Still it feels like a very Pango-speficic problem.

Reply:
I could be convinced that pango should just take languages as-is. That will probably be less of a concern if language does not affect font selection ordering anymore.

Reply:
> But is this necessary?

For using fontconfig properly it's necessary to use this API, because of fonts like DejaVu which eg. have Arabic glyphs but are not deemed suitable for Persian (low quality). So I made this complex machinery in Pango, using this API, such that if you set LANGUAGES=en:fa, then any Arabic-script text will use the preferred font for 'fa'.

Without the API proposed here that would not be possible, because I need to know that Arabic script is not used to write en, to pass fa to fontconfig.

If I pass both en,fa in that order to fontconfig, it will return DejaVu, which is what I want to avoid.

Reply:
I still feel it is a shoot yourself in the foot-kind API but I understand there is a need and I don’t want to block you, so feel free to finish and merge this as you see fit.